reve/reve/lib/ModuleSimplifier.cpp

```diff
@@ -355,6 +360,166 @@ llvm::PreservedAnalyses RemoveLifetimeCallsPass::run(
+        for(llvm::Use &U : Fun.uses()) {
+            // Figure out whether the return value is used after each call
+            
+            if(auto CI = llvm::dyn_cast<llvm::CallInst>(U.getUser())) {
+                if(CI->getCalledFunction() != &Fun)
+                    can_replace = false; // Different function is called, Fun is an argument
+#ifdef DEBUG
+                CI->print(llvm::errs(), false); llvm::errs() << "\n";
+                for(llvm::Use &UU : CI->uses()) {
+                    llvm::errs() << "  "; UU.getUser()->print(llvm::errs(), false); llvm::errs() << "\n";
+                }
+#endif
+                if(!CI->use_empty())
+                    can_replace = false; // The return value is actually used
+            } else if(auto II = llvm::dyn_cast<llvm::InvokeInst>(U.getUser())) {
+                if(II->getCalledFunction() != &Fun)
+                    can_replace = false; // Different function is called, Fun is an argument
+#ifdef DEBUG
+                II->print(llvm::errs(), false); llvm::errs() << "\n";
+                for(llvm::Use &UU : II->uses()) {
+                    llvm::errs() << "  "; UU.getUser()->print(llvm::errs(), false); llvm::errs() << "\n";
+                }
+#endif
+                if(!II->use_empty()) // The return value is actually used
+                    can_replace = false;
+            } else can_replace = false; // The function is used somewhere as an argument, therefore
+                                        // it ought not to be replaced
+        }
+        
+        if(can_replace) {  
+            // Create the header of the new function
+            std::vector<llvm::Type*> FAT_New (Fun.getFunctionType()->param_begin(), Fun.getFunctionType()->param_end());
+            llvm::FunctionType *FT_New = llvm::FunctionType::get(llvm::Type::getVoidTy(Fun.getContext()), FAT_New, Fun.isVarArg());
+            llvm::Function *Fun_New = llvm::Function::Create(FT_New, Fun.getLinkage(), Fun.getName(), Fun.getParent());
+            
+            // Copy the attributes from the old function and delete the ones related to the return value
+            Fun_New->copyAttributesFrom(&Fun);
+            for(llvm::Attribute::AttrKind AK : badAttributes) {
+                Fun_New->removeAttribute(llvm::AttributeList::ReturnIndex, AK);
+                Fun_New->removeAttribute(llvm::AttributeList::FunctionIndex, AK);
+            }
+            Fun_New->takeName(&Fun);
+            
+            for(llvm::Function::arg_iterator AI = Fun.arg_begin(), AE = Fun.arg_end(), NAI = Fun_New->arg_begin();
+                AI != AE; 
+                ++AI, ++NAI)
+            {
+                NAI->takeName(AI); // Set the names of all arguments of the new function
+            }
+            
+            Fun_New->getBasicBlockList().splice(Fun_New->begin(), Fun.getBasicBlockList()); // Copy the function body
+            
+            // Replace return instructions on ends of basic blocks with ret void
```

Comment:
Please, add note that this is currently not used since we apply the pass to declarations only. Same applies 2 lines above.